### ISSUE TITLE
This loop is a prime candidate for paralleization

### ISSUE THREAD

**Issue by jfiscus** [usnistgov/ActEV_Scorer#31]
https://github.com/usnistgov/ActEV_Scorer/blob/a18cc25d22424975770a908ab9a3f0d0d79e908a/lib/protocols/default.py#L86

**Comment by Chocorean** [comment]
Yeah that is what I try to parallelize. More precisely:
```python3
        for activity, activity_properties in self.activity_index.items():
            refs = ref_by_act.get(activity, [])
            syss = sys_by_act.get(activity, [])
            kernel = kernel_builder(activity, activity_properties, refs, syss)
            for rs, ss in cohort_gen(refs, syss):
                c, m, f = perform_alignment(rs, ss, kernel)  # <-- issue here for multiprocessing

                alignment_recs.extend(c)
                alignment_recs.extend(m)
                alignment_recs.extend(f)
            
            i += 1
            sys_nbr += len(syss)
```
Let me explain what the issues I met:
- `threading` lib : The python interpreter we use (CPython) prevent us to use multithreading efficiently, because of [GIL](https://wiki.python.org/moin/GlobalInterpreterLock). Basically, only one thread can use the CPU because of GIL. It works great with I/O bounded threads but not in our case (CPU bounded threads).
- `multiprocessing` lib : Processes don't share memory, so I need to pass arguments such as *activities* and the *kernel* function to the new process. The issue here is when passing args to a process, they are serialized with `pickle`, which can only serialize objects in the global scope. Plus as we need to reload everything, memory usage will <strike>probably</strike> increase.

Right now:
- `threading` could work if we use another python3 interpreter which doesn't have a GIL. Unfortunately, none is available atm (cf IronPython, Jython, etc)
- I have more hopes for `multiprocessing`; I still have a few ideas I'd like to try (e.g. try to declare dynamically variable in globals). 
- Workaround: Maybe we could separate this piece of code in another file which would be called through shell (GIL wouldn't block us as it would be a different process I guess). We could also dump arguments into files, and read them in the thread, but that sounds dirty. I'm still trying to come up with more elegant ideas. Doin a lot of research. Maybe change the language only for computing alignments to get rid of GIL and python3 drawbacks?

**Comment by Chocorean** [comment]
Fixed by #33
